File: BWT using C++/BWT.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

struct RotationIndex {
    int index;
    string rotation;
};

bool compareSuffix(const RotationIndex& a, const RotationIndex& b) {
    return a.rotation < b.rotation;
}

string burrowsWheelerTransform(const string& text) {
    int n = text.size();
    vector<RotationIndex> rotations(n);

    // Generate indices of cyclic rotations
    for (int i = 0; i < n; ++i) {
        rotations[i].index = i;
        rotations[i].rotation = text.substr(i) + text.substr(0, i);
    }

    // Sort the indices based on the corresponding cyclic rotations using Suffix Array
    sort(rotations.begin(), rotations.end(), compareSuffix);

    // Extract the last characters of each rotation
    string transformedText;
    for (int i = 0; i < n; ++i) {
        transformedText.push_back(rotations[i].rotation[n - 1]);
    }

    return transformedText;
}

int main() {
    int choice;
    cout << "Choose input option:" << endl;
    cout << "1. Enter text manually" << endl;
    cout << "2. Input text from file" << endl;
    cin >> choice;
    string text;
    if (choice == 1) {
        cout << "Enter the text: ";
        cin.ignore(); // Ignore the newline character left by previous cin
        getline(cin, text);
    } else if (choice == 2) {
        string filename;
        cout << "Enter the filename containing the text: ";
        cin >> filename;
        ifstream file(filename.c_str());
        if (!file) {
            cerr << "Error: Unable to open file." << endl;
            return 1;
        }
        string line;
        while (getline(file, line)) {
            text += line;
        }
    } else {
        cerr << "Invalid choice." << endl;
        return 1;
    }
    string bwt = burrowsWheelerTransform(text);
    cout << "Burrows-Wheeler Transform: " << bwt << endl;
    return 0;
}

File: BWT using python/BWT.py
def burrows_wheeler_transform(text):
    n = len(text)
    rotations = [text[i:] + text[:i] for i in range(n)]
    sorted_rotations = sorted(enumerate(rotations), key=lambda x: x[1])
    transformed_text = ''.join(rotation[-1] for _, rotation in sorted_rotations)
    return transformed_text

def main():
    choice = int(input("Choose input option:\n1. Enter text manually\n2. Input text from file\n"))
    if choice == 1:
        text = input("Enter the text: ")
    elif choice == 2:
        filename = input("Enter the filename containing the text: ")
        with open(filename, 'r') as file:
            text = ''.join(file.readlines())
    else:
        print("Invalid choice.")
        return

    bwt = burrows_wheeler_transform(text)
    print("Burrows-Wheeler Transform:", bwt)

if __name__ == "__main__":
    main()
